# 2024 Homework

## Q.1 - Pandas version

In [78]:
import pandas as pd
import numpy as np

pd.__version__

'2.3.1'

## Getting the data

In [35]:
# Read in the data
file_location = "https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv"
laptop_data = pd.read_csv(
    file_location)

laptop_data.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [36]:
laptop_data.describe()

RAM      Storage       Screen  Final Price
count  2160.000000  2160.000000  2156.000000  2160.000000
mean     15.413889   596.294444    15.168112  1312.638509
std       9.867815   361.220506     1.203329   911.475417
min       4.000000     0.000000    10.100000   201.050000
25%       8.000000   256.000000    14.000000   661.082500
50%      16.000000   512.000000    15.600000  1031.945000
75%      16.000000  1000.000000    15.600000  1708.970000
max     128.000000  4000.000000    18.000000  7150.470000

## Q.2 - Records Count

In [85]:
print(len(laptop_data))

print(laptop_data.shape[0])
# Should be the same answer

2160
2160


## Q.3 - Laptop brands

In [6]:
# To count unique values, use nunique() function

# Convert strings to lower
laptop_brands = laptop_data.Brand.str.lower()

laptop_brands.nunique()



27

In [7]:
laptop_data.Brand.nunique()

# Both return 27 so there must be 27 unique brands

27

In [8]:
# To see the unique values: 
laptop_data.Brand.unique()

array(['Asus', 'Alurin', 'MSI', 'HP', 'Lenovo', 'Medion', 'Acer', 'Apple',
       'Razer', 'Gigabyte', 'Dell', 'LG', 'Samsung', 'PcCom', 'Microsoft',
       'Primux', 'Prixton', 'Dynabook Toshiba', 'Thomson', 'Denver',
       'Deep Gaming', 'Vant', 'Innjoo', 'Jetwing', 'Millenium', 'Realme',
       'Toshiba'], dtype=object)

## Q.4 - Missing Values

In [16]:
nulls = laptop_data.isnull().sum()
nulls

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

In [19]:
col_contains_nulls = nulls > 0
col_contains_nulls

Laptop          False
Status          False
Brand           False
Model           False
CPU             False
RAM             False
Storage         False
Storage type     True
GPU              True
Screen           True
Touch           False
Final Price     False
dtype: bool

In [21]:
col_contains_nulls.sum()

np.int64(3)

Three columns contain nulls

## Q.5 - Maximum Final Price

In [25]:
dell_laptops = laptop_data[laptop_data.Brand == "Dell"]
dell_laptops

Laptop       Status Brand  \
161   Dell Vostro 3510 Intel Core i3-1115G4/8 GB/256...          New  Dell   
271   Dell Vostro 3510 Intel Core i5-1135G7/8 GB/256...          New  Dell   
281   Dell Vostro 3520 Intel Core i7-1255U/16GB/512G...          New  Dell   
319   Dell Latitude 3520 Intel Core i5-1135G7/8GB/25...          New  Dell   
328   Dell Vostro 3510 Intel Core i5-1135G7/16 GB/51...          New  Dell   
...                                                 ...          ...   ...   
1843  Dell Vostro 5301 Intel Core i5-1135G7/8GB/256G...  Refurbished  Dell   
1844  Dell Vostro 5301 Intel Core i5-1135G7/8GB/512G...  Refurbished  Dell   
1845  Dell Vostro 5502 Intel Core i5-1135G7/8GB/512G...  Refurbished  Dell   
1846  Dell XPS 13 9300 Intel Core i5-1035G1/8GB/512G...  Refurbished  Dell   
1847  Dell XPS 7390 Intel Core i7-10510U/16GB/512GB ...  Refurbished  Dell   

         Model            CPU  RAM  Storage Storage type  GPU  Screen Touch  \
161     Vostro  Intel Core i3    8      256          SSD  NaN    15.6    No   
271     Vostro  Intel Core i5    8      256          SSD  NaN    15.6    No   
281     Vostro  Intel Core i7   16      512          SSD  NaN    15.6    No   
319   Latitude  Intel Core i5    8      256          SSD  NaN    15.6    No   
328     Vostro  Intel Core i5   16      512          SSD  NaN    15.6    No   
...        ...            ...  ...      ...          ...  ...     ...   ...   
1843    Vostro  Intel Core i5    8      256          SSD  NaN    13.3    No   
1844    Vostro  Intel Core i5    8      512          SSD  NaN    13.3    No   
1845    Vostro  Intel Core i5    8      512          SSD  NaN    15.6    No   
1846       XPS  Intel Core i5    8      512          SSD  NaN    13.4    No   
1847       XPS  Intel Core i7   16      512          SSD  NaN    13.3    No   

      Final Price  
161        599.00  
271        690.15  
281        889.00  
319        789.00  
328        869.00  
...           ...  
1843       420.74  
1844       511.59  
1845       549.93  
1846       699.00  
1847      1086.88  

[84 rows x 12 columns]

In [27]:
dell_laptops["Final Price"].max()

np.float64(3936.0)

In [87]:
laptop_data.groupby("Brand")["Final Price"].agg(["min", "mean", "max"])

min         mean      max
Brand                                          
Acer               264.14  1001.285766  3691.00
Alurin             239.00   484.701379   869.00
Apple              299.00  1578.227672  3849.00
Asus               239.25  1269.380699  5758.14
Deep Gaming       1334.00  1505.377500  1639.01
Dell               379.00  1153.839881  3936.00
Denver             329.95   329.950000   329.95
Dynabook Toshiba   397.29   999.197895  1805.01
Gigabyte           799.00  1698.488958  3799.00
HP                 210.14   952.628478  5368.77
Innjoo             251.40   347.888333   431.38
Jetwing            469.27   469.270000   469.27
LG                 609.04  1730.028438  2399.00
Lenovo             239.00  1087.723880  5018.14
MSI                444.19  1913.451006  7150.47
Medion             246.89   905.331875  3799.00
Microsoft          599.01  1595.543247  3747.91
Millenium         1640.98  1976.845000  2312.71
PcCom             1049.00  1606.762917  1949.90
Primux             208.99   339.618750   599.41
Prixton            246.95   292.280000   329.95
Razer             1232.74  3265.701622  4999.01
Realme             999.00   999.000000   999.00
Samsung            365.90  1595.511364  3699.01
Thomson            201.05   316.860000   436.56
Toshiba            799.00   799.000000   799.00
Vant               632.00  1047.455455  1217.01

Maximum price of a dell laptop is 3936

## Q.6 - Median Value of Screen

### 6.1 - Mediuan value of all screens

In [44]:
screens = laptop_data.Screen
screens.describe()

count    2156.000000
mean       15.168112
std         1.203329
min        10.100000
25%        14.000000
50%        15.600000
75%        15.600000
max        18.000000
Name: Screen, dtype: float64

In [54]:
median_screen_value = screens.median()
median_screen_value # 15,6

np.float64(15.6)

### 6.2 - Modal value of all screens

In [38]:
modal_screen_value = screens.mode()
modal_screen_value

0    15.6
Name: Screen, dtype: float64

In [59]:
screens_filled = screens.fillna(15.6)
screens_filled.describe()

count    2160.000000
mean       15.168912
std         1.202357
min        10.100000
25%        14.000000
50%        15.600000
75%        15.600000
max        18.000000
Name: Screen, dtype: float64

In [56]:
new_mean_value = screens_filled.median()

In [60]:
new_mean_value / median_screen_value

np.float64(1.0)

## Q.7 - Sum of weights

In [62]:
innjoo_laptops = laptop_data[laptop_data.Brand == "Innjoo"]
innjoo_laptops

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

In [64]:
innjoo_data = innjoo_laptops[["RAM", "Storage", "Screen"]]
innjoo_data

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [66]:
# Convert the data to a np array, just .array
X = innjoo_data.values
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [74]:
XT = X.transpose()
XT

array([[  8. ,   8. ,   4. ,   6. ,   6. ,   6. ],
       [256. , 512. ,  64. ,  64. , 128. , 128. ],
       [ 15.6,  15.6,  14.1,  14.1,  14.1,  14.1]])

In [76]:
XTX = XT.dot(X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [80]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [81]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])

In [83]:
w = (XTX_inv.dot(XT)).dot(y)
w

array([45.58076606,  0.42783519, 45.29127938])

In [84]:
w.sum()

np.float64(91.2998806299555)

In [ ]:
# Sum is 91.3

In [88]:
print(XT.dot(X))
print(XT @ X)

[[2.52000e+02 8.32000e+03 5.59800e+02]
 [8.32000e+03 3.68640e+05 1.73952e+04]
 [5.59800e+02 1.73952e+04 1.28196e+03]]
[[2.52000e+02 8.32000e+03 5.59800e+02]
 [8.32000e+03 3.68640e+05 1.73952e+04]
 [5.59800e+02 1.73952e+04 1.28196e+03]]
